cside-main/src/system_manager.c

```diff
@@ -1,14 +1,151 @@
+//
+// Copyright (c) 2022 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+
+#include <esp_event.h>
+#include <esp_log.h>
+#include <esp_system.h>
+#include <esp_wifi.h>
+#include <freertos/FreeRTOS.h>
+#include <freertos/event_groups.h>
+#include <freertos/task.h>
+#include <nvs_flash.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <zenoh-pico.h>
 #include "system_manager.h"
 
 
+// WiFi credentials
+/* TODO: how should we set the wifi credentials? */
+#define ESP_WIFI_SSID "SET SSID HERE"
+#define ESP_WIFI_PASS "SET PASSWORD HERE"
+
+/// WiFi event handler
+#define ESP_MAXIMUM_RETRY 5
+#define WIFI_CONNECTED_BIT BIT0
+#define SHOULD_SKIP_WIFI_INIT false
+
+static bool s_is_wifi_connected = false;
+static EventGroupHandle_t s_event_group_handler;
+static int s_retry_count = 0;
+static bool _system_initialized = false;
+
+/* PRIVATE FUNCTIONS */
+static void event_handler(void *arg, esp_event_base_t event_base, int32_t event_id, void *event_data) {
+    if (event_base == WIFI_EVENT && event_id == WIFI_EVENT_STA_START) {
+        esp_wifi_connect();
+    } else if (event_base == WIFI_EVENT && event_id == WIFI_EVENT_STA_DISCONNECTED) {
+        if (s_retry_count < ESP_MAXIMUM_RETRY) {
+            esp_wifi_connect();
+            s_retry_count++;
+        }
+    } else if (event_base == IP_EVENT && event_id == IP_EVENT_STA_GOT_IP) {
+        xEventGroupSetBits(s_event_group_handler, WIFI_CONNECTED_BIT);
+        s_retry_count = 0;
+    }
+}
+
+
+void _system_manager_board_init(system_manager_t* system_manager)
+{
+    esp_err_t ret = nvs_flash_init();
+    if (ret == ESP_ERR_NVS_NO_FREE_PAGES || ret == ESP_ERR_NVS_NEW_VERSION_FOUND) {
+        ESP_ERROR_CHECK(nvs_flash_erase());
+        ret = nvs_flash_init();
+    }
+    ESP_ERROR_CHECK(ret);
+
+    int *context = (int *)malloc(sizeof(int));
```

Comment:
Should remove these two lines since they serve no purpose (and create memory leak)

Reply:
are you talking about these lines?
    ```
int *context = (int *)malloc(sizeof(int));
    *context = 0;
```
or are you also including ESP_ERROR_CHECK(ret)?